### PR TITLE
fix(rds): auto minor upgrade is enabled even if the engine's version is pinned

### LINKED ISSUE #15475: (aws-rds): DatabaseCluster should default `auto_minor_version_upgrade` to `False` if engine version is pinned
This may not be _truly_ a bug, but it is not an ideal setup.

When using `aws_rds.DatabaseCluster`, I had a failing over-the-top deploy yesterday because the Aurora PostgreSQL version had been upgraded outside of the CFN stack. I was unaware that this was a behaviour, especially as I had specifically targeted a version of PostgreSQL.

I eventually realised there is the flag `auto_minor_version_upgrade` which I'd not set, but I noticed that its default value is `True`.

What I think would be more user friendly is if this value defaults to `False` **if** `engine` has a version specified.

### Reproduction Steps

```python
from aws_cdk import aws_rds

# Specify an RDS Cluster with a Version set like so
aws_rds.DatabaseCluster(
    self,
    id="rds-cluster",
    engine=aws_rds.DatabaseClusterEngine.aurora_postgres(
        version=aws_rds.AuroraPostgresEngineVersion.VER_10_12
    ),
    ...
)
...
```

### What did you expect to happen?

Deploying the stack **without** further changes to the  `DatabaseCluster` construct would result in a successful deploy.

### What actually happened?

The deploy failed with the following error message:
```
23/42 |4:47:39 PM | UPDATE_FAILED        | AWS::RDS::DBInstance                        | prod-downloader-rds/Instance2 (proddownloaderrdsInstance2BE42B470) The specified DB Instance is a member of a cluster. Modify the DB engine version for the DB Cluster using the ModifyDbCluster API (Service: AmazonRDS; Status Code: 400; Error Code: InvalidParameterCombination; Request ID: fe271c1d-16c8-4acf-928e-dd93574d7ac4; Proxy: null)
	createInstances (/tmp/jsii-kernel-0VxvEJ/node_modules/@aws-cdk/aws-rds/lib/cluster.js:360:26)
	\_ new DatabaseCluster (/tmp/jsii-kernel-0VxvEJ/node_modules/@aws-cdk/aws-rds/lib/cluster.js:201:34)
	\_ /tmp/tmpbizvy7oq/lib/program.js:8154:58
	\_ Kernel._wrapSandboxCode (/tmp/tmpbizvy7oq/lib/program.js:8582:24)
	\_ Kernel._create (/tmp/tmpbizvy7oq/lib/program.js:8154:34)
	\_ Kernel.create (/tmp/tmpbizvy7oq/lib/program.js:7895:29)
	\_ KernelHost.processRequest (/tmp/tmpbizvy7oq/lib/program.js:9479:36)
	\_ KernelHost.run (/tmp/tmpbizvy7oq/lib/program.js:9442:22)
	\_ Immediate._onImmediate (/tmp/tmpbizvy7oq/lib/program.js:9443:46)
	\_ processImmediate (internal/timers.js:461:21)
```

When I checked the RDS Console, I could indeed see that even though my Stack was deployed with `10_12` as the version, it had been upgraded from underneath me to `10_14`, breaking my deployment.

### Environment

  - **CDK CLI Version: 1.105.0 (build 4813992)**
  - **Framework Version: 1.105.0**
  - **Node.js Version: v12.22.1** 
  - **OS: Ubuntu 20.04.2**
  - **Language (Version): Python 3.8.6** 

### Other

Following on from a discussion with @skinny85 here https://cdk-dev.slack.com/archives/C018XT6REKT/p1625763866267400

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by nohack**
Closes #15475

### CODE REVIEW

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 1]
```diff
@@ -718,6 +718,11 @@ function createInstances(cluster: DatabaseClusterNew, props: DatabaseClusterBase
       props.clusterIdentifier != null ? `${props.clusterIdentifier}instance${instanceIndex}` :
         undefined;
 
+    // if engine version is pinned and autoMinorVersionUpgrade is not passed in props, we set this property to false
+    let autoMinorVersionUpgrade = props.instanceProps.autoMinorVersionUpgrade;
+    if (props.engine.engineVersion !== undefined && props.instanceProps.autoMinorVersionUpgrade === undefined) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:723]: Check `engine.engineVersion.fullVersion !=== undefined` instead.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 2]
```diff
@@ -738,7 +743,7 @@ function createInstances(cluster: DatabaseClusterNew, props: DatabaseClusterBase
       dbParameterGroupName: instanceParameterGroupConfig?.parameterGroupName,
       monitoringInterval: props.monitoringInterval && props.monitoringInterval.toSeconds(),
       monitoringRoleArn: monitoringRole && monitoringRole.roleArn,
-      autoMinorVersionUpgrade: props.instanceProps.autoMinorVersionUpgrade,
+      autoMinorVersionUpgrade: autoMinorVersionUpgrade,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:746]: Let's get rid of the `autoMinorVersionUpgrade` variable, and just do things inline:

```suggestion
      autoMinorVersionUpgrade: props.engine.engineVersion?.fullVersion && props.instanceProps.autoMinorVersionUpgrade === undefined
        // if engine version is pinned and autoMinorVersionUpgrade is not passed in props, we set this property to false
        ? false
        : props.instanceProps.autoMinorVersionUpgrade,
```

File: packages/@aws-cdk/aws-rds/lib/instance.ts [hunk 3]
```diff
@@ -872,11 +871,18 @@ abstract class DatabaseInstanceSource extends DatabaseInstanceNew implements IDa
...
+    if (props.engine.engineVersion !== undefined && props.autoMinorVersionUpgrade === undefined) {
+      autoMinorVersionUpgrade = false;
+    }
+
     this.instanceType = props.instanceType ?? ec2.InstanceType.of(ec2.InstanceClass.M5, ec2.InstanceSize.LARGE);
 
     const instanceParameterGroupConfig = props.parameterGroup?.bindToInstance({});
     this.sourceCfnProps = {
       ...this.newCfnProps,
+      autoMinorVersionUpgrade,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:885]: Why did you move this here?

File: packages/@aws-cdk/aws-rds/lib/instance.ts [hunk 4]
```diff
@@ -872,11 +871,18 @@ abstract class DatabaseInstanceSource extends DatabaseInstanceNew implements IDa
       }
     }
 
+    // if engine version is pinned and autoMinorVersionUpgrade is not passed in props, we set this property to false
+    let autoMinorVersionUpgrade = props.autoMinorVersionUpgrade;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:875]: Same comments here:

1. Get rid of the `autoMinorVersionUpgrade` variable, and do this work inline.
2. Check `props.engine.engineVersion?.fullVersion`.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 5]
```diff
@@ -432,6 +432,47 @@ describe('cluster', () => {
     });
   });
 
+  test('cluster with engine version and if no autoMinorVersionUpgrade in props disable automatic upgrade of minor version', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:435]:
```suggestion
  test('cluster with an engine in a specific version and without autoMinorVersionUpgrade set disables automatic minor version upgrades', () => {
```

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 6]
```diff
@@ -432,6 +432,47 @@ describe('cluster', () => {
...
+        vpc,
+      },
+    });
+
+    expect(stack).toHaveResource('AWS::RDS::DBInstance', {
+      AutoMinorVersionUpgrade: false,
+    });
+  });
+
+  test('cluster with engine version and if autoMinorVersionUpgrade allow automatic upgrade of minor version', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:455]:
```suggestion
  test("cluster with an engine in a specific version and with autoMinorVersionUpgrade set to 'true' enables automatic minor version upgrades", () => {
```

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 7]
```diff
@@ -1706,6 +1747,45 @@ describe('cluster', () => {
     });
   });
 
+  test('cluster from snapshot with engine version and if no autoMinorVersionUpgrade in props disable automatic upgrade of minor version', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:1750]:
```suggestion
  test('cluster from snapshot with an engine in a specific version and without autoMinorVersionUpgrade set disables automatic minor version upgrades', () => {

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 8]
```diff
@@ -1706,6 +1747,45 @@ describe('cluster', () => {
...
+      snapshotIdentifier: 'mySnapshot',
+    });
+
+    // THEN
+    expect(stack).toHaveResourceLike('AWS::RDS::DBInstance', {
+      AutoMinorVersionUpgrade: false,
+    });
+  });
+
+  test('cluster from snapshot with engine version and if autoMinorVersionUpgrade in props enable automatic upgrade of minor version', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:1769]:
```suggestion
  test("cluster from snapshot with an engine in a specific version and with autoMinorVersionUpgrade set to 'true' enables automatic minor version upgrades", () => {
```

File: packages/@aws-cdk/aws-rds/test/instance.test.ts [hunk 9]
```diff
@@ -284,6 +284,33 @@ describe('instance', () => {
 
   });
 
+  test('instance with engine version and if no autoMinorVersionUpgrade in props disable automatic upgrade of minor version', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:287]:
```suggestion
  test('instance with an engine in a specific version and without autoMinorVersionUpgrade set disables automatic minor version upgrades', () => {
```

File: packages/@aws-cdk/aws-rds/test/instance.test.ts [hunk 10]
```diff
@@ -284,6 +284,33 @@ describe('instance', () => {
...
+      }),
+      vpc,
+    });
+
+    expect(stack).toHaveResource('AWS::RDS::DBInstance', {
+      AutoMinorVersionUpgrade: false,
+    });
+  });
+
+  test('instance with engine version and if autoMinorVersionUpgrade enable automatic upgrade of minor version', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:300]:
```suggestion
  test("instance with an engine in a specific version and with autoMinorVersionUpgrade set to 'true' enables automatic minor version upgrades", () => {
```

File: packages/@aws-cdk/aws-rds/test/instance.test.ts [hunk 11]
```diff
@@ -401,6 +428,31 @@ describe('instance', () => {
 
 
     });
+
+    test('instance from snapshot with engine version and if no autoMinorVersionUpgrade in props disable automatic upgrade of minor version', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:432]:
```suggestion
  test('instance from snapshot with an engine in a specific version and without autoMinorVersionUpgrade set disables automatic minor version upgrades', () => {
```

File: packages/@aws-cdk/aws-rds/test/instance.test.ts [hunk 12]
```diff
@@ -401,6 +428,31 @@ describe('instance', () => {
...
+        engine: rds.DatabaseInstanceEngine.mysql({ version: rds.MysqlEngineVersion.VER_8_0_19 }),
+        vpc,
+      });
+
+      expect(stack).toHaveResourceLike('AWS::RDS::DBInstance', {
+        AutoMinorVersionUpgrade: false,
+      });
+    });
+
+    test('instance from snapshot with engine version and autoMinorVersionUpgrade enable automatic upgrade of minor version', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:444]:
```suggestion
  test("instance from snapshot with an engine in a specific version and with autoMinorVersionUpgrade set to 'true' enables automatic minor version upgrades", () => {
```

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks great @nohack, thanks so much for the contribution!

A few minor comments.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts (hunk 1 above)
Comment (nohack) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts]: Sure.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts (hunk 2 above)
Comment (nohack) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts]: Done.thanks.

File: packages/@aws-cdk/aws-rds/lib/instance.ts (hunk 3 above)
Comment (nohack) [reply packages/@aws-cdk/aws-rds/lib/instance.ts]: autoMinorVersionUpgrade is present in DatabaseInstanceNewProps and engine is present in DatabaseInstanceSourceProps.
So I deferred setting autoMinorVersionUpgrade and updated it in DatabaseInstanceSource constructor.

**Comment by skinny85** [comment]
@nohack please re-request my review (there's a button in the top-right of the PR window, next to my avatar) once you're ready for another round of reviews.

**Review by skinny85 — DISMISSED** [review]
Looks fantastic @nohack, thanks so much for the contribution!

**Comment by skinny85** [comment]
@nohack looks like you might have to update some integration tests:

```
@aws-cdk/aws-codepipeline-actions: ================================================================================
@aws-cdk/aws-codepipeline-actions: Test Suites: 17 passed, 17 total
@aws-cdk/aws-codepipeline-actions: Tests:       109 passed, 109 total
@aws-cdk/aws-codepipeline-actions: Snapshots:   0 total
@aws-cdk/aws-codepipeline-actions: Time:        14.201 s
@aws-cdk/aws-codepipeline-actions: Ran all test suites.
@aws-cdk/aws-rds: Verifying integ.proxy.js against integ.proxy.expected.json ... CHANGED.
@aws-cdk/aws-rds: Resources
@aws-cdk/aws-rds: [~] AWS::RDS::DBInstance dbInstance4076B1EC may be replaced
@aws-cdk/aws-rds:  â””â”€ [+] AutoMinorVersionUpgrade (may cause replacement)
@aws-cdk/aws-rds:      â””â”€ false
@aws-cdk/aws-ecs-patterns: Verifying ec2/integ.scheduled-ecs-task.lit.js against ec2/integ.scheduled-ecs-task.lit.expected.json ... OK.
@aws-cdk/aws-rds: Verifying integ.serverless-cluster.js against integ.serverless-cluster.expected.json ... OK.
@aws-cdk/aws-rds: Error: Some stacks have changed. To verify that they still deploy successfully, run: 'yarn integ integ.instance-from-generated-password.js integ.instance-s3.js integ.proxy.js'
```

Instructions for how to do that are in our ["Contributing" guide](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#integration-tests).

**Review by skinny85 — CHANGES_REQUESTED** [review]
Putting the PR in "request changes" until the integ tests are updated.

**Comment by nohack** [comment]
sure.will update.

**Comment by nohack** [comment]
I get the below error on running integ test cases.This happens on master as well.
The password is auto generated and stored in secrets manager right.
`  
yarn integ integ.instance-from-generated-password.js integ.instance-s3.js integ.proxy.js

..
CREATE_FAILED        | AWS::RDS::DBInstance                        | Instance (....) The parameter MasterUserPassword is not a valid password. Only printable ASCII characters besides '/', '@', '"', ' ' may be used. (Service: AmazonRDS; Status Code: 400; Error Code: InvalidParameterValue; Request ID: .....; Proxy: null)
	new DatabaseInstance (/home/../workspace/aws-cdk/packages/@aws-cdk/aws-rds/lib/instance.js:386:26)
	\_ new DatabaseInstanceStack (/home/../workspace/aws-cdk/packages/@aws-cdk/aws-rds/test/integ.instance-from-generated-password.js:11:9)
	\_ Object.<anonymous> (/home/../workspace/aws-cdk/packages/@aws-cdk/aws-rds/test/integ.instance-from-generated-password.js:23:1)
`

**Comment by skinny85** [comment]
> I get the below error on running integ test cases.This happens on master as well.
> The password is auto generated and stored in secrets manager right.
> 
> ```
> yarn integ integ.instance-from-generated-password.js integ.instance-s3.js integ.proxy.js
> 
> CREATE_FAILED        | AWS::RDS::DBInstance                        | Instance (....) The parameter MasterUserPassword is not a valid password. Only printable ASCII characters besides '/', '@', '"', ' ' may be used. (Service: AmazonRDS; Status Code: 400; Error Code: InvalidParameterValue; Request ID: .....; Proxy: null)
> 	new DatabaseInstance (/home/../workspace/aws-cdk/packages/@aws-cdk/aws-rds/lib/instance.js:386:26)
> 	\_ new DatabaseInstanceStack (/home/../workspace/aws-cdk/packages/@aws-cdk/aws-rds/test/integ.instance-from-generated-password.js:11:9)
> 	\_ Object.<anonymous> (/home/../workspace/aws-cdk/packages/@aws-cdk/aws-rds/test/integ.instance-from-generated-password.js:23:1)
> ```

Hmmm. Perhaps we need to update [this line](https://github.com/aws/aws-cdk/blob/cdee1af03c34a1c08988e672bae6edc2538a8877/packages/%40aws-cdk/aws-rds/test/integ.instance-from-generated-password.ts#L16-L16) to include all of the characters mentioned in that error message (`/`, `@`, `"`, and ` `)? From what I can see, only `@` is present there right now.
